proof/hash_to_g1.saw

```diff
@@ -0,0 +1,688 @@
+swu_impl_on_curve <- simp_then_admit {{ \u -> is_point_projective Curve_E' (map_to_curve_simple_swu_impl u) == True }};
+
+// point_dadd implements add:
+dadd_thm <- admit_cryptol (rewrite (cryptol_ss ()) {{ \ p1 p2 -> affinify Curve_E' (point_dadd Fp p1 p2 A' False) == add Curve_E' (affinify Curve_E' p1) (affinify Curve_E' p2) }});
+
+// the addition of two points on the curve is on the curve:
+point_dadd_on_curve <- simp_then_admit {{ \p q -> is_point_projective Curve_E' (point_dadd Fp p q A' False) == if is_point_projective Curve_E' p /\ is_point_projective Curve_E' q then True else apply is_point_projective Curve_E' (point_dadd Fp p q A' False) }};
+
+// adding a point to its double can be done with POINTonE1_add (cheaper than dadd) because the double of a point is never equal to the point unless the point is point_O:
+add_double_lemma <- admit_cryptol {{ \p -> affinify E (point_add Fp (point_double Fp p) p) == add E (affinify E (point_double Fp p)) (affinify E p) }};
+
+// We will need to prove that some expression involving field Fp operations are
+// equivalent.  To do so, we will try to rewrite them to a normal form with the
+// following rules. The problem is that we cannot use commutativity or
+// associativity, which would make the rewriter loop....
+
+zero_unit_rewrites <- for [ // TODO: here we are cheating: this is true only if we use the (Z p) type
+    {{ \p -> Fp.add (Fp.field_zero, p) == p }}
+  , {{ \p -> Fp.add (p, Fp.field_zero) == p }}
+  , {{ \p -> Fp.mul (p, Fp.field_unit) == p }}
+  , {{ \p -> Fp.mul (Fp.field_unit, p) == p }}
+  , {{ \p -> Fp.mul (p, Fp.field_zero) == Fp.field_zero }}
+  , {{ \p -> Fp.mul (Fp.field_zero, p) == Fp.field_zero }}
+  , {{ \p -> Fp.div (p, Fp.field_zero) == Fp.field_zero }}
+  , {{ \p -> Fp.div (p, Fp.field_unit) == Fp.field_unit }}
+  , {{ \p -> Fp.div (Fp.field_zero, p) == Fp.field_zero }}
+] simp_then_admit;
+
+// NOTE: those rules should be sound since we have `Fp.div (x,0) == 0`
+push_div_out_rewrites <- for [
+    {{ \x y a b -> Fp.mul (Fp.div (x,y), Fp.div (a,b)) == Fp.div (Fp.mul (x,a), Fp.mul (y,b)) }}
+  , {{ \x y z -> Fp.mul (Fp.div (x,y), z) == Fp.div (Fp.mul (x,z), y) }}
+  , {{ \x y z -> Fp.mul (z, Fp.div (x,y)) == Fp.div (Fp.mul (x,z), y) }}
+  , {{ \x y z -> Fp.div (Fp.div (z, x), y) == Fp.div (z, Fp.mul (x,y)) }}
+] simp_then_admit;
+
+sum_of_prods_rewrites <- for [
+  {{ \a b c -> Fp.mul (Fp.add (a, b), c) == Fp.add (Fp.mul (a,c), Fp.mul (b,c)) }}
+  , {{ \a b c -> Fp.mul (c, Fp.add (a, b)) == Fp.add (Fp.mul (a,c), Fp.mul (b,c)) }}
+] simp_then_admit;
+
+unsquare_rewrite <- simp_then_admit {{ \a -> Fp.sq a == Fp.mul (a,a) }};
+
+left_fold_mul <- simp_then_admit {{ \x y z -> Fp.mul (x, Fp.mul (y,z)) == Fp.mul (Fp.mul (x, y),z) }};
+
+// this one is not included in algebra_simpset below:
+right_fold_mul <- simp_then_admit {{ \x y z -> Fp.mul (Fp.mul (x, y),z) == Fp.mul (x, Fp.mul (y,z)) }};
+
+simp_div_rewrites <- for [
```

Comment:
I don't think any of these are true when x is 0.

Reply:
Yes, I need to fix that